### DISCUSSION TITLE
Dependencies

### DISCUSSION THREAD

**Discussion by gherbstman** [Yevrag35/DsRegModule#6] [Q&A]
It does seen this module has a dependency on the system having the DsRegCmd.exe or something else system version related. It does not run on a 2012 R2 server.

Can you enlighten me? Maybe the dependencies section can be updated.

**Comment by Yevrag35** [comment]
So, technically, Windows Server 2012 R2 is considered a "down-level" device.  And because of that, dsregcmd.exe is not included in the OS install.  But while you can't retrieve the status of a down-level device with this module, you could run it to pull the information from Windows 10 or Server 2016+ machines remotely from one.

I'll think about how I can add that verbiage to the README or something.

Here's a link to a Microsoft article on how to retrieve Azure AD registration status on down-level devices: https://docs.microsoft.com/en-us/azure/active-directory/devices/troubleshoot-hybrid-join-windows-legacy

EDIT - Yes, this module is a 'wrapper' and, as such, it does not include the program that it is intended to wrap.

**Comment by gherbstman** [comment]
Thanks for the update.

As I am running scripts locally through an rmm tool I don't really want the complexity of anything remote. New paragraph I did come up with a workaround for my scripts in that I check for the existence of DNS command and if it does not exist I just use other means to get the local domain. I also assume at that point it cannot be azure ad joined because it is as you say a down level device.

The code is simple and only a few lines of code. If you want to incorporate that I'd be glad to send you what I have.
________________________________

**Reply by Yevrag35** [reply]
I created an [issue](https://github.com/Yevrag35/DsRegModule/issues/7), and I'll see about adding a custom error to the commands if the OS version does not have the dsregcmd executable.

**Comment by gherbstman** [comment]
If you find this useful.

Function Get-CommandExists ([string]$cmd) {
    return ((Get-Command $cmd -ErrorAction SilentlyContinue) -ne $null)
}

# Older systems do not support getting the status in the modern way.
If (-Not (Get-CommandExists("dsregcmd.exe"))) {
    Write-Output "DSRegCmd missing, Trying alternate method"

    $domainName = (Get-CIMInstance Win32_NTDomain).DomainName
    If ($domainName -ne $null) {
        $StatusText = "Domain Joined: " + $domainName
    }
     else {
        $StatusText = "Domain Unkown"
     }
}

**Comment by Yevrag35** [comment]
Curiously, if all you're trying to pull is the Domain Name, why not just use <code>(Get-CimInstance Win32_NTDomain).DomainName</code> for all situations?  

Are you using the module for other things too?

**Comment by gherbstman** [comment]
We are not just wanting the AD info. We also want to know if the device is AAD joined and report the name of AAD. This is just a small snipit. On computers that do not have DSRegCmd, it cannot be AAD joined anyway.

**Comment by Yevrag35** [comment]
You can, actually, Azure AD Hybrid Join down-level devices.  However, it's quite a bit harder to do so as the machine would have needed to download a specific program called "autoworkplace.exe" from Microsoft.  

If you have not enabled down-level device within your organization's Azure AD Connect installation, then yes, you can use your method just fine.

However, If you're an MSP, you won't be able to assume that all down-level devices are not Azure AD Hybrid Joined as the customer _may_ have enabled down-level support (unless you're 100% sure they wouldn't have done that).

A better way, in my opinion, would be a check like this:

```powershell
Function Test-DsReg() {
   Test-Path -Path "$env:WINDIR\System32\dsregcmd.exe" -PathType Leaf
}

Function Test-Autoworkplace() {
   Test-Path -Path "$env:PROGRAMFILES\Microsoft Workplace Join\autoworkplace.exe" -PathType Leaf
}

if (-not (Test-DsReg)) {
   if (-not (Test-Autoworkplace)) {
      # Machine is down-level and 100% not Hybrid Azure AD Joined
   }
   else {
      # Machine is down-level and possibly could be Hybrid Azure AD Joined
   }
}
```

**Comment by gherbstman** [comment]
Thanks Mike, that was great info. We have only been AAD joining W10 computers. You are right, I would want to cover older computers. And I learned something.

Thanks.
